[taufiq] iP

taufiq opened the pull request:
# Dukio
> Do what you want – Sun Tze 🔥

Dukio is the unknown brother of Mario, Nintendo's favourite character. It's
- Fast
- Amazing!
- Swag 😎

All you need to do is:
1. Download from [here](https://github.com/taufiq/ip/releases/tag/v0.1)
2. Double-click it
3. Start managing your tasks **_TODAY_**!

Features:
- [X] Manage Tasks
- [X] Manage Deadlines
- [X] Manage Events
- [X] Store Tasks!
- [ ] Predict the future!

Start getting used to the bot saying:
```
Mamma-mia!
```

Everytime you do something it doesn't understand 😎

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

taufiq committed:
feat: Add custom intro

Added my first custom intro to the bot!

taufiq committed:
feat: add custom intro

taufiq committed:
feat: Add echo

The bot can now repeat whatever the user inputs.

taufiq committed:
feat: add echo

taufiq committed:
feat: Add add,list command

The bot can now add tasks and list them.

taufiq committed:
feat: add add,list command

taufiq committed:
feat: add marking as done

taufiq committed:
feat: Add marking as done

Tasks now can be marked as done and their status is also shown when listed.

taufiq committed:
refactor+feat: Add todo,event,deadline + refactor

Add 3 variants of Tasks which all inherit from tasks. I also introduced OOP as I noticed a lot of duplicated code that's all in one file. I basically created a Parser that parses user input to ouptut commands, a State class to hold the state of the app, and variants of Commands to execute the user's intentions.

taufiq committed:
refactor+feat: add todo,event,deadline + refactor

taufiq committed:
refactor: Swap out to use OOP + error handling

Added OOP abstraction and error handling for 2 types of errors (Incorrect argument and Unknown commands)

taufiq committed:
refactor: swap out to use OOP + error handling

Added OOP abstraction and error handling for 2 types of errors (Incorrect argument and Unknown commands)

taufiq committed:
feat: Add delete ability

Users can now delete tasks.

taufiq committed:
feat: add delete ability

taufiq committed:
feat: add test cases

taufiq committed:
feat: Add test cases

Added input test cases and expected output so that the bot is tested reliably.

taufiq committed:
feat: Introduce persistence

I added persistence by using a csv-esque format. Of course, I didn't have the interest to take care of malicious inputs where the task description is the same as the delimiter used (`|`) so I just outright banned it.

taufiq committed:
feat: introduce persistence

I added persistence by using a csv-esque format. Of course, I didn't have the interest to take care of malicious inputs where the task description is the same as the delimiter used (`|`) so I just outright banned it.

taufiq committed:
feat: Parse dates for tasks

Users can now input date strings into their tasks. If an invalid format is added, a parsing exception is thrown to reflect to the user.

taufiq committed:
feat: parse dates for tasks

Users can now input date strings into their tasks. If an invalid format is added, a parsing exception is thrown to reflect to the user.

taufiq committed:
refactor: Create ui class

ui class to handle ui stuff.

taufiq committed:
refactor: create ui class

ui class to handle ui stuff.

taufiq committed:
refactor: move all files into `duke` package

taufiq committed:
refactor: Move all files into `duke` package

taufiq committed:
Merge remote-tracking branch 'origin/add-gradle-support'

taufiq committed:
fix: Downgrade jdk to match gradle merge

I was using Java 21 and when I downgraded I was denied the syntactic sugar privileges that 21 gave :sad

taufiq committed:
fix: downgrade jdk to match gradle merge

I was using Java 21 and when I downgraded I was denied the syntactic sugar privileges that 21 gave :sad

Lin-Shuang-Shuang commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,22 @@
+        LIST,
+        DELETE,
+        ADD,
+        MARK
```
I like that you used enumerations.

Lin-Shuang-Shuang commented on src/main/java/duke/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,22 @@
+package duke;
+
+import java.time.LocalDate;
+
```
Maybe you could write javadocs for your main files?

Lin-Shuang-Shuang commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,94 @@
+import java.util.List;
+
+public class Parser {
+    public static Command parse(String text, State state) throws DukeException, DateTimeParseException {
```
Perhaps you could break down your parse method into smaller methods like parseAddCommand etc.?

Lin-Shuang-Shuang commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,7 @@
+package duke;
+
+public class Ui {
```
Should your Ui do more than just print out Strings?

taufiq replied:
yeah i'll throw in some input or sth

Lin-Shuang-Shuang reviewed (commented):
LGTM. Overall, your code is easy to read and understand. Perhaps your Ui can be better implemented?

taufiq committed:
test: Add ParserTest test cases

Added a unit test for the `Parser` class to ensure that the Parser has all it's edge cases covered.

taufiq committed:
test: add ParserTest test cases

taufiq committed:
test: Add StateTest test cases

Added unit tests for State class to make sure State class works as expected.

taufiq committed:
test: add StateTest test cases

taufiq committed:
docs: add documentation

taufiq committed:
docs: Add documentation

Added documentation so that developers who work on this project can navigate around the project.

J-wang-CSMA commented on src/main/java/duke/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,22 @@
+
+import java.time.LocalDate;
+
+public class AddDeadlineCommand extends Command {
```
Perhaps some javadocs could be written for the methods and fields such that anyone who reads it can understand the code changes instantly 😄

J-wang-CSMA commented on src/main/java/duke/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,22 @@
+    @Override
+    public void execute(State state, Ui ui) {
+        Task newDeadline = new Deadline(
+                getText(),
```
I like the linebreaks here! Makes the code very legible.

J-wang-CSMA commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,21 @@
+
+public class Deadline extends Task {
+    LocalDate byWhen;
+    public Deadline(String description, LocalDate byWhen, boolean done) {
```
Might want to consider naming the boolean "done" something that sounds like a boolean according to the Java coding standard provided. For example, we can name the boolean variable "isDone" instead of "done". :)

taufiq replied:
Yep, will add that in. Thanks for catching this :)

J-wang-CSMA commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,28 @@
+import java.time.format.DateTimeParseException;
+import java.util.Scanner;
+
+public class Duke {
```
Very nice adherance to the bracket coding style requirements in the main function. Good job!

J-wang-CSMA commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,95 @@
+        }
+        String header = parts.get(0);
+        Command command;
+        switch(header) {
```
Perhaps we can follow the Java coding standard provided on the CS2103/T website when it comes to switch statements? (The switch statement should have the following form: Note there is no indentation for case clauses.)
```` java
switch (condition) {
case ABC:
    statements;
    // Fallthrough
case DEF:
    statements;
    break;
case XYZ:
    statements;
    break;
default:
    statements;
    break;
}

taufiq replied:
Thanks for the catch for this as well :)

J-wang-CSMA commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,89 @@
+package duke;
+
+import java.io.*;
```
Perhaps you can import functions or classes explcitly in line with the CS2103/T coding style requirements instead of using a wildcard here?

taufiq replied:
Have changed it

J-wang-CSMA commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,89 @@
+import java.util.Scanner;
+
+public class Storage {
+    enum TaskType {
```
I like the ENUM usage here.

J-wang-CSMA commented on src/main/java/duke/Util.java:
```diff
@@ -0,0 +1,15 @@
+package duke;
+
+public class Util {
```
Some javadocs here would be nice to tell others what the Util class is for.

J-wang-CSMA reviewed (commented):
Looks good to me, mostly just needs some more javadocs comments to tell others what the code is doing. Other than that just some minor slip-ups. Overall very good adherance to the CS2103/T coding standard.

taufiq committed:
Merge pull request #1 from taufiq/branch-A-JavaDoc

docs: add documentation

taufiq committed:
Merge pull request #2 from taufiq/branch-A-CodingStandard

chore: Adhere to coding standards

taufiq committed:
feat: Add find command

Now users can find tasks by searching for its description with a `find <text>` command.

taufiq committed:
Merge pull request #3 from taufiq/branch-Level-9

feat: Add find command

taufiq committed:
chore: Remove wildcard import for Storage

Adhere to the norms!

taufiq committed:
style: Rename boolean to adhere to standards

I named some of my booleans as `done` but it should have a prefix like `is`/`was`/etc. I've made the necessary changes.

taufiq committed:
style: Remove indentation in switch statements

Unindented the `case`s in the switch statement.

taufiq committed:
build: Include JavaFX dependencies

taufiq committed:
feat: Add FXML files and resources

taufiq committed:
feat: Add JavaFX GUI

For this I had to refactor and return a `String` when executing commands so that I can pass it to the `MainWindow` controller to output the text when communicating.

Other than that, I've added in the primary work of the GUI and it's working fully :fire

taufiq committed:
docs: Add in missing return comments for execute

I didn't include what the `execute` command returns in the JavaDocs.

taufiq committed:
Merge pull request #4 from taufiq/branch-Level-10

feat: Add JavaFX GUI

taufiq committed:
feat: Add assertions in codebase

Currently, there are no assertions within the codebase, which is not ideal as there are points where we do make assumptions about how the code behaves.

To correct this, Assertions are added in places deemed fit to make such assertions. For example, in the `DeleteCommand`, we assert that the index is within bounds before executing a delete. Another example would be list where the StringBuilder that we're returning does not return an empty output (even when there's no tasks).

These are fair assertions to make.

taufiq committed:
refactor: remove `CommandType` Enum

The `CommandType` Enum was being used to label the nature of commands executed.

However, these enums were not being used anywhere in the code other than to label the command. It remains a persistent issue as after going through many iterations, this enum marking was not being used anywhere in the codebase, which signifies the presence of dead code.

Removing the Enum completely allows us to reduce the codebase size whilst maintaining good meaning within the codebase for other developers.

The whole enum is wiped as it is not necessary to remain in the codebase.

taufiq committed:
Merge pull request #6 from taufiq/branch-A-CodeQuality

refactor: remove `CommandType` Enum

taufiq committed:
Merge branch 'master' into branch-A-Assertions

taufiq committed:
Merge pull request #7 from taufiq/branch-A-Assertions

feat: Add assertions

taufiq committed:
feat: Allow users to view schedule for a day

This commit adds a `view_schedule <date>` command that allows users to view the Deadlines and Events that fall under the specified date.

taufiq committed:
refactor: rename to Dukio

taufiq committed:
docs: Add User Guide

taufiq committed:
refactor: update old names

taufiq committed:
build: Update build information
